### PR TITLE
dm/syncer: fix directly use types.NewDatum

### PR DESCRIPTION
**Pull request by WizardXiao**
### What problem does this PR solve? 

fix https://github.com/pingcap/ticdc/issues/3550
### What is changed and how it works?

use `AdjustBinaryProtocolForDatum` instead of `types.NewDatum`
### Check List 

Tests 

 - Unit test

### Release note 

```release-note
None
```

### CODE REVIEW

File: dm/syncer/dml.go [hunk 1]
```diff
@@ -686,14 +687,18 @@ func genKeyList(table string, columns []*model.ColumnInfo, dataSeq []interface{}
 }
 
 // truncateIndexValues truncate prefix index from data.
-func truncateIndexValues(indexColumns *model.IndexInfo, tiColumns []*model.ColumnInfo, data []interface{}) []interface{} {
+func truncateIndexValues(ctx sessionctx.Context, indexColumns *model.IndexInfo, tiColumns []*model.ColumnInfo, data []interface{}) []interface{} {
 	values := make([]interface{}, 0, len(indexColumns.Columns))
+	datums, err := utils.AdjustBinaryProtocolForDatum(ctx, data, tiColumns)
```
Comment (Ehco1996) [inline dm/syncer/dml.go:692]: seems we can use `tablecodec.TruncateIndexValues` to save a for loop below

**Review by Ehco1996 — COMMENTED** [review]
rest LGTM

File: dm/syncer/dml.go (hunk 1 above)
Comment (WizardXiao) [reply dm/syncer/dml.go:692]: yes, `AdjustBinaryProtocolForDatum` which is diffrent from `types.NewDatum` returns `[]types.Datum` , i will fix.

**Review by Ehco1996 — APPROVED** [review]

**Review by lance6716 — APPROVED** [review]

**Pull request merged**